packages/validator/src/validate.decorator.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/validator
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {JSONSchema6} from 'json-schema';
+import * as AJV from 'ajv';
+import {
+  MetadataAccessor,
+  MetadataInspector,
+  ParameterDecoratorFactory,
+} from '@loopback/metadata';
+import * as HttpErrors from 'http-errors';
+
+const debug = require('debug')('loopback:validator');
+
+// tslint:disable-next-line:no-any
+export type Validator = (arg: any) => boolean | Promise<boolean>;
+
+export const VALIDATION_KEY = MetadataAccessor.create<JSONSchema6 | Validator>(
+  'validation.parameter',
+);
+
+export function validatable() {
+  return function(
+    target: Object,
+    member: string,
+    // tslint:disable-next-line:no-any
+    descriptor: TypedPropertyDescriptor<(...args: any[]) => any>,
+  ) {
+    const originalMethod = descriptor.value;
```

Comment:
I am not a fan of decorators that are directly modifying properties and methods of the decorated objects, I think the complexity of handling edge cases (like the number of arguments stored in `.length` property) is not worth the benefits.

What are the arguments for using this approach, as opposed to letting the REST layer to perform the actual validation based on metadata provided by `@param` and/or other metadata-only decorators?

Reply:
My main point for using this approach is for the clean error tracing. Our sequence makes the error tracing irrelevant, but I personally thought having the arguments validated right as the controller method is invoked seemed like it just made sense at high level.

Another argument for it is that the decorators would be automatically compatible with any future server extensions. This approach would let extension developers implement validation for their servers just by having the users import the package.

As for the edge cases, the only ones I see aside from `length` are `name` and `toString()`. Are there any that I may have missed?

Reply:
I see your points.

To me, replacing controller methods seems too magic, I prefer to have more control. As I was envisioning validation, different server extensions will call the validation step in their transport-specific sequence and each server extension will decide when is the right time to perform the validation and coercion.


Few things to consider:

Does it make sense to split coercion and validation into two steps? Are these steps really independent? Consider the REST sequence where `parseParams` action is responsible for extracting typed arguments from the HTTP request (and thus coerce the values from string sources like query string values to correct types). If the validation happens inside `invoke` action, then how are we going to treat invalid input values? Are we going to perform a partial validation as part of coercion, and then another validation as part of invoking controller method?

Let's say a controller method accepts an argument of type "unsigned 32bit integer" and the caller provided a string (e.g. `foo`) or a value that's out of bounds (e.g. `-1` or `17e10`).  Using the default sequence (see below), what should `parseParams` return for this input argument? https://github.com/strongloop/loopback-next/blob/43383f5ff8ab98880e5a3bdb91d91c03e10c5ba5/packages/rest/src/sequence.ts#L106-L108

In my opinion, both coercion and validation should happen at the same time.